File: packages/isl-runtime-go/constraints.go
package islruntime

import (
	"encoding/json"
	"fmt"
	"os"
	"regexp"
	"strings"
)

// BehaviorConstraint represents constraints for a behavior
type BehaviorConstraint struct {
	Name          string   `json:"name"`
	Preconditions []string `json:"preconditions"`
	Postconditions []string `json:"postconditions"`
	Invariants    []string `json:"invariants"`
}

// DomainConstraints represents constraints for a domain
type DomainConstraints struct {
	Domain          string               `json:"domain"`
	Behaviors       []BehaviorConstraint  `json:"behaviors"`
	GlobalInvariants []string            `json:"global_invariants"`
}

// ConstraintLoader loads ISL constraints from files or compiled JSON
type ConstraintLoader struct{}

// NewConstraintLoader creates a new constraint loader
func NewConstraintLoader() *ConstraintLoader {
	return &ConstraintLoader{}
}

// LoadFromFile loads constraints from an ISL spec file
//
// Note: This is a simplified parser. For full ISL parsing, use the TypeScript parser
// and export to JSON format, then use LoadFromJSON.
func (cl *ConstraintLoader) LoadFromFile(path string) (*DomainConstraints, error) {
	content, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("failed to read file: %w", err)
	}
	return cl.parseISL(string(content))
}

// LoadFromJSON loads constraints from compiled JSON format
func (cl *ConstraintLoader) LoadFromJSON(path string) (*DomainConstraints, error) {
	content, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("failed to read file: %w", err)
	}
	var constraints DomainConstraints
	if err := json.Unmarshal(content, &constraints); err != nil {
		return nil, fmt.Errorf("failed to parse JSON: %w", err)
	}
	return &constraints, nil
}

func (cl *ConstraintLoader) parseISL(content string) (*DomainConstraints, error) {
	domain := "Unknown"
	var behaviors []BehaviorConstraint
	var globalInvariants []string

	lines := strings.Split(content, "\n")
	inBehavior := false
	var currentBehavior *BehaviorConstraint

	for _, line := range lines {
		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, "domain ") {
			parts := strings.Fields(line)
			if len(parts) > 1 {
				domain = parts[1]
			}
		} else if strings.HasPrefix(line, "behavior ") {
			if currentBehavior != nil {
				behaviors = append(behaviors, *currentBehavior)
			}
			parts := strings.Fields(line)
			name := ""
			if len(parts) > 1 {
				name = parts[1]
			}
			currentBehavior = &BehaviorConstraint{
				Name:          name,
				Preconditions: []string{},
				Postconditions: []string{},
				Invariants:    []string{},
			}
			inBehavior = true
		} else if inBehavior && currentBehavior != nil {
			if strings.Contains(line, "precondition") || strings.Contains(line, "pre ") {
				if expr := cl.extractExpression(line); expr != "" {
					currentBehavior.Preconditions = append(currentBehavior.Preconditions, expr)
				}
			} else if strings.Contains(line, "postcondition") || strings.Contains(line, "post ") {
				if expr := cl.extractExpression(line); expr != "" {
					currentBehavior.Postconditions = append(currentBehavior.Postconditions, expr)
				}
			} else if strings.Contains(line, "invariant") {
				if expr := cl.extractExpression(line); expr != "" {
					currentBehavior.Invariants = append(currentBehavior.Invariants, expr)
				}
			}
			if line == "}" {
				inBehavior = false
			}
		} else if strings.HasPrefix(line, "invariant ") {
			if expr := cl.extractExpression(line); expr != "" {
				globalInvariants = append(globalInvariants, expr)
			}
		}
	}

	if currentBehavior != nil {
		behaviors = append(behaviors, *currentBehavior)
	}

	if domain == "" {
		domain = "Unknown"
	}

	return &DomainConstraints{
		Domain:          domain,
		Behaviors:       behaviors,
		GlobalInvariants: globalInvariants,
	}, nil
}

func (cl *ConstraintLoader) extractExpression(line string) string {
	keywords := []string{"precondition", "postcondition", "invariant", "pre ", "post "}
	for _, keyword := range keywords {
		if idx := strings.Index(line, keyword); idx != -1 {
			expr := strings.TrimSpace(line[idx+len(keyword):])
			if expr != "" && !strings.HasPrefix(expr, "{") {
				return expr
			}
		}
	}
	return ""
}

File: packages/isl-runtime-rs/src/constraints.rs
//! ISL constraint loader

use crate::types::*;
use serde_json::Value;
use std::fs;
use std::path::Path;

/// Loads ISL constraints from files or compiled JSON
pub struct ConstraintLoader;

impl ConstraintLoader {
    /// Create a new constraint loader
    pub fn new() -> Self {
        Self
    }

    /// Load constraints from an ISL spec file
    ///
    /// Note: This is a simplified parser. For full ISL parsing, use the TypeScript parser
    /// and export to JSON format, then use `load_from_json`.
    pub fn load_from_file(&self, path: impl AsRef<Path>) -> Result<DomainConstraints, Box<dyn std::error::Error>> {
        let content = fs::read_to_string(path)?;
        self.parse_isl(&content)
    }

    /// Load constraints from compiled JSON format
    pub fn load_from_json(&self, path: impl AsRef<Path>) -> Result<DomainConstraints, Box<dyn std::error::Error>> {
        let content = fs::read_to_string(path)?;
        let constraints: DomainConstraints = serde_json::from_str(&content)?;
        Ok(constraints)
    }

    /// Parse ISL spec (simplified parser)
    ///
    /// For production use, compile ISL files to JSON using the TypeScript tooling,
    /// then use `load_from_json`.
    fn parse_isl(&self, content: &str) -> Result<DomainConstraints, Box<dyn std::error::Error>> {
        let mut domain = String::new();
        let mut behaviors = Vec::new();
        let mut global_invariants = Vec::new();

        // Extract domain name
        for line in content.lines() {
            let line = line.trim();
            if line.starts_with("domain ") {
                domain = line
                    .replace("domain", "")
                    .trim()
                    .split_whitespace()
                    .next()
                    .unwrap_or("Unknown")
                    .to_string();
                break;
            }
        }

        // Extract behaviors (simplified - for production use JSON export)
        let mut in_behavior = false;
        let mut current_behavior: Option<BehaviorConstraint> = None;

        for line in content.lines() {
            let line = line.trim();
            if line.starts_with("behavior ") {
                if let Some(behavior) = current_behavior.take() {
                    behaviors.push(behavior);
                }
                let name = line
                    .replace("behavior", "")
                    .trim()
                    .split_whitespace()
                    .next()
                    .unwrap_or("")
                    .to_string();
                current_behavior = Some(BehaviorConstraint {
                    name,
                    preconditions: Vec::new(),
                    postconditions: Vec::new(),
                    invariants: Vec::new(),
                });
                in_behavior = true;
            } else if in_behavior {
                if let Some(ref mut behavior) = current_behavior {
                    if line.contains("precondition") || line.contains("pre ") {
                        // Extract precondition expression
                        if let Some(expr) = self.extract_expression(line) {
                            behavior.preconditions.push(expr);
                        }
                    } else if line.contains("postcondition") || line.contains("post ") {
                        // Extract postcondition expression
                        if let Some(expr) = self.extract_expression(line) {
                            behavior.postconditions.push(expr);
                        }
                    } else if line.contains("invariant") {
                        // Extract invariant expression
                        if let Some(expr) = self.extract_expression(line) {
                            behavior.invariants.push(expr);
                        }
                    }
                }
                if line == "}" {
                    in_behavior = false;
                }
            } else if line.starts_with("invariant ") {
                if let Some(expr) = self.extract_expression(line) {
                    global_invariants.push(expr);
                }
            }
        }

        if let Some(behavior) = current_behavior {
            behaviors.push(behavior);
        }

        Ok(DomainConstraints {
            domain: if domain.is_empty() { "Unknown".to_string() } else { domain },
            behaviors,
            global_invariants,
        })
    }

    fn extract_expression(&self, line: &str) -> Option<String> {
        // Simple extraction - look for content after keywords
        let keywords = ["precondition", "postcondition", "invariant", "pre ", "post "];
        for keyword in &keywords {
            if let Some(pos) = line.find(keyword) {
                let expr = line[pos + keyword.len()..].trim();
                if !expr.is_empty() && !expr.starts_with('{') {
                    return Some(expr.to_string());
                }
            }
        }
        None
    }
}

impl Default for ConstraintLoader {
    fn default() -> Self {
        Self::new()
    }
}
